### PR TITLE
Fixes #1649. Exclusion rules and other cached resolution fixes

### LINKED ISSUE #1649: Cached resolution does not respect excludes
In cached resolution, minigraph keys do not consider excluded dependencies.  If one project excludes a transitive dependency, and another does not, it becomes non-deterministic which minigraph will be used under cached resolution.

Minimal repro:
- Clone https://github.com/rossabaker/sbt-cached-resolution-exclusion
- Run `sbt clean test`

Expectation:
- Both specs should pass: two projects should be able to independently prune their transitive dependencies.

Result:
- One will not.  For me it's ExcludeSpec, but I suspect this may be non-deterministic.

Real world motivation:
- hadoop-core brings in old servlet-apis under the `org.mortbay.jetty` organization.  This is not evicted by the Java Servlet API from `javax.servlet`, resulting in two conflicting versions of `javax.servlet` on the classpath.  If the user has compiled another Hadoop project that does not exclude the servlet-api, the cache is polluted for the project that does need to exclude it.

Extra credit:
- Works with `excludeAll`
- Works with `ivyXML`

### PR DESCRIPTION
**Pull request by eed3si9n**
Fixes #1649.

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @eed3si9n,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.typesafe.com/contribute/cla](http://www.typesafe.com/contribute/cla)

File: ivy/src/main/scala/sbt/ivyint/CachedResolutionResolveEngine.scala [hunk 1]
```diff
@@ -59,11 +59,28 @@ private[sbt] class CachedResolutionResolveCache() {
     }
   def buildArtificialModuleDescriptor(dd: DependencyDescriptor, rootModuleConfigs: Vector[IvyConfiguration], prOpt: Option[ProjectResolver]): (DefaultModuleDescriptor, Boolean) =
     {
+      def excludeRuleString(rule: ExcludeRule): String =
+        s"""Exclude(${rule.getId},${rule.getConfigurations.mkString(",")},${rule.getMatcher})"""
+      def includeRuleString(rule: IncludeRule): String =
+        s"""Include(${rule.getId},${rule.getConfigurations.mkString(",")},${rule.getMatcher})"""
```
Comment (jsuereth) [inline ivy/src/main/scala/sbt/ivyint/CachedResolutionResolveEngine.scala:68]: Hmm..... this feels less consistent....   Are the toStrings of rule.getMatcher pretty consistent in general?

**Comment by jsuereth** [comment]
Generally LGTM

File: ivy/src/main/scala/sbt/ivyint/CachedResolutionResolveEngine.scala (hunk 1 above)
Comment (eed3si9n) [inline ivy/src/main/scala/sbt/ivyint/CachedResolutionResolveEngine.scala:68]: Yes. Ivy, like many Java libraries, have nice `toString()` normally. :neutral_face:

**Comment by eed3si9n** [comment]
Since this was not merged yet, I'm stacking another bug fix on top.

**Comment by jsuereth** [comment]
LGTM, again

**Comment by eed3si9n** [comment]
More fixes were added yesterday. Tests failing, but not sure if it's me or Travis.

**Comment by eed3si9n** [comment]
@jsuereth so you were right about Travis error being legit. Good job, Travis/scripted!
`updateClassifiers` sets a flag called `missingOk`, which allows Ivy to explorer around to grab JARs, but not let the unresolved error propagate as error. I needed to port the behavior for cached resolution.

**Comment by jsuereth** [comment]
LGTM

**Pull request merged**